frame/gmetadata/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+        ) -> DispatchResultWithPostInfo {
+            Self::check_namespace(origin, key.ns)?;
+            let mut old_value = IndexStore::<T>::get(&key);
+            match &mut old_value {
+                Some(old_value) => match old_value.data.iter().position(|v| v == &value) {
+                    Some(idx) => {
+                        old_value.data.remove(idx);
+                        old_value.data.sort();
+                        old_value.update_time =
+                            T::UnixTime::now().as_millis().saturated_into::<u64>();
+                        IndexStore::<T>::insert(key.clone(), old_value);
+                    }
+                    None => {}
+                },
+                None => {}
+            }
+            Self::deposit_event(Event::StateUpdate(
+                req_id,
+                key.ns,
+                key.table.clone().into(),
+                key.pk.clone().into(),
+            ));
+            Self::deposit_event(Event::IndexUpdate(
+                req_id,
+                key.ns,
+                key.table.into(),
+                key.pk.into(),
+            ));
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn value(key: GmetadataKey) -> Option<Vec<u8>> {
+            match ValueStore::<T>::get(key) {
+                Some(value) => Some(value.data),
+                None => None,
+            }
+        }
+
+        fn check_owner_or_root(
+            origin: &OriginFor<T>,
+            owners: &Vec<T::AccountId>,
+        ) -> DispatchResult {
+            match ensure_signed(origin.clone()) {
+                Ok(who) => {
+                    if !owners.contains(&who) {
+                        return Err(DispatchError::BadOrigin);
+                    }
+                }
+                Err(_) => ensure_root(origin.clone())?,
+            }
+            Ok(().into())
+        }
+
+        fn check_namespace(origin: OriginFor<T>, ns_id: u32) -> DispatchResultWithPostInfo {
+            let ns = match NamespaceStore::<T>::get(ns_id) {
+                Some(ns) => ns,
+                None => return Err(Error::<T>::NamespaceNotFound.into()),
+            };
+            Self::check_owner_or_root(&origin, &ns.owners)?;
+            Ok(().into())
+        }
+
+        fn check_key(key: &GmetadataKey) -> DispatchResultWithPostInfo {
+            if key.table.len() > 100 {
+                return Err(Error::<T>::InvalidKey.into());
+            }
+            if key.pk.len() > 500 {
+                return Err(Error::<T>::InvalidKey.into());
+            }
+            Ok(().into())
+        }
+
+        pub fn query_with_index(
+            index_key: GmetadataKey,
+            value_key: GmetadataKey,
+            start: HexBytes,
+            limit: u64,
+        ) -> GmetadataQueryResult {
+            Self::query_with_indexes(vec![index_key], value_key, start, limit)
+        }
+
+        pub fn query_with_indexes(
```

Comment:
I'm not so sure about this. enumerating unbounded lists that can be manipulated by the user on chain can lead to DoS attacks